File: frontend/src/lib/server/v2/goals.test.js
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';

const apiRequest = vi.fn();
vi.mock('$lib/api-helpers.js', () => ({ apiRequest: (...a) => apiRequest(...a) }));
vi.mock('./org-people.js', () => ({
  getOrgPeopleAndTeams: async () => ({ people: [], teams: [] })
}));

const { listGoals } = await import('./goals.js');

/** A JWT whose payload carries just the role claim `viewerRole` reads. */
function tokenFor(role) {
  const payload = Buffer.from(JSON.stringify({ role })).toString('base64url');
  return `header.${payload}.signature`;
}

const cookies = /** @type {any} */ ({ get: () => tokenFor('ADMIN') });

/** One `SalesGoalSerializer` row, with only the fields under test spelled out. */
function goal(over = {}) {
  return {
    id: 'g1',
    name: 'Goal',
    goal_type: 'REVENUE',
    target_value: '100',
    period_type: 'MONTHLY',
    period_start: '2026-01-01',
    period_end: '2026-12-31',
    is_active: true,
    progress_value: 10,
    progress_percent: 10,
    status: 'on_track',
    ...over
  };
}

function respond({ goals = [], leaderboard = [] }) {
  apiRequest.mockImplementation(async (url) =>
    url.includes('leaderboard') ? { leaderboard } : { goals }
  );
}

describe('listGoals totals', () => {
  beforeEach(() => {
    apiRequest.mockReset();
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('sums targets and attainment over the active goals only', async () => {
    vi.setSystemTime(new Date(2026, 5, 1, 12));
    respond({
      goals: [
        goal({ id: 'a', target_value: '100', progress_value: 50 }),
        goal({ id: 'b', target_value: '200', progress_value: 20 }),
        goal({ id: 'c', target_value: '999', progress_value: 999, is_active: false })
      ]
    });

    const { totals } = await listGoals({ cookies });
    expect(totals.count).toBe(3);
    expect(totals.active).toBe(2);
    expect(totals.target).toBe(300);
    expect(totals.achieved).toBe(70);
  });

  it('still counts a goal whose period ends today as behind pace', async () => {
    // The boundary this fixes. `new Date('2026-06-01')` is midnight UTC, so
    // measuring it against `Date.now()` dropped the goal out of this count
    // part-way through its own final day, for every timezone east of
    // Greenwich and for most of the day. Noon local is chosen deliberately:
    // it is past midnight UTC, which is exactly when the old code broke.
    vi.setSystemTime(new Date(2026, 5, 1, 12));
    respond({ goals: [goal({ status: 'behind', period_end: '2026-06-01' })] });

    const { totals } = await listGoals({ cookies });
    expect(totals.behind).toBe(1);
  });

  it('leaves out a goal whose period has already ended', async () => {
    vi.setSystemTime(new Date(2026, 5, 1, 12));
    respond({ goals: [goal({ status: 'behind', period_end: '2026-05-31' })] });

    const { totals } = await listGoals({ cookies });
    expect(totals.behind).toBe(0);
  });

  it('leaves out a retired goal even when it is behind and still open', async () => {
    vi.setSystemTime(new Date(2026, 5, 1, 12));
    respond({
      goals: [goal({ status: 'behind', period_end: '2026-12-31', is_active: false })]
    });

    const { totals } = await listGoals({ cookies });
    expect(totals.behind).toBe(0);
  });
});

describe('listGoals leaderboard', () => {
  beforeEach(() => {
    apiRequest.mockReset();
  });

  it('reads the name the endpoint sends, which is no longer an email', async () => {
    // The row used to carry the address twice, as `name` and as `email`, so
    // this list printed raw addresses. There is nothing to fall back to now.
    respond({
      leaderboard: [
        {
          rank: 1,
          goal_id: 'g1',
          user: { id: 'p1', name: 'Ada Lovelace' },
          target: 100,
          achieved: 104,
          percent: 104
        }
      ]
    });

    const { leaderboard } = await listGoals({ cookies });
    expect(leaderboard[0].user).toBe('Ada Lovelace');
    // Uncapped, unlike `progress_percent`, which the model caps at 100.
    expect(leaderboard[0].percent).toBe(104);
  });

  it('says Unknown rather than blank when the user block is missing', async () => {
    respond({ leaderboard: [{ rank: 1, goal_id: 'g1' }] });
    const { leaderboard } = await listGoals({ cookies });
    expect(leaderboard[0].user).toBe('Unknown');
  });

  it('is empty rather than throwing when the endpoint narrows a member to nothing', async () => {
    // An ordinary outcome now that the board is scoped like the list: somebody
    // with no current monthly goal of their own ranks against nobody.
    respond({ leaderboard: [] });
    const { leaderboard } = await listGoals({ cookies });
    expect(leaderboard).toEqual([]);
  });
});

describe('listGoals can_edit', () => {
  beforeEach(() => {
    apiRequest.mockReset();
    respond({});
  });

  it('is true for an admin and false for a member', async () => {
    const asAdmin = await listGoals({ cookies });
    expect(asAdmin.can_edit).toBe(true);

    const asMember = await listGoals({
      cookies: /** @type {any} */ ({ get: () => tokenFor('USER') })
    });
    expect(asMember.can_edit).toBe(false);
  });

  it('is false when there is no token, rather than throwing', async () => {
    const signedOut = await listGoals({
      cookies: /** @type {any} */ ({ get: () => undefined })
    });
    expect(signedOut.can_edit).toBe(false);
  });
});

describe('activity goals and deal type weights', () => {
  beforeEach(() => {
    apiRequest.mockReset();
  });

  it('offers ACTIVITIES as a goal type the backend accepts', async () => {
    const { GOAL_TYPES } = await import('./goals.js');
    expect(GOAL_TYPES).toContain('ACTIVITIES');
  });

  it('names the five deal types a weight can be set on', async () => {
    const { DEAL_TYPES } = await import('./goals.js');
    expect(DEAL_TYPES).toEqual([
      'NEW_BUSINESS',
      'EXISTING_BUSINESS',
      'RENEWAL',
      'UPSELL',
      'CROSS_SELL'
    ]);
  });

  it('reads a weight map off the form, skipping the types left blank', async () => {
    const { weightsFromForm } = await import('./goals.js');
    const form = new FormData();
    form.set('weight_RENEWAL', '0.5');
    form.set('weight_NEW_BUSINESS', '');
    form.set('weight_UPSELL', '  ');

    expect(weightsFromForm(form)).toEqual({ RENEWAL: 0.5 });
  });

  it('passes a non-numeric weight through so the backend can name it', async () => {
    const { weightsFromForm } = await import('./goals.js');
    const form = new FormData();
    form.set('weight_RENEWAL', 'heavy');

    expect(weightsFromForm(form)).toEqual({ RENEWAL: 'heavy' });
  });

  it('sends the weight map on create', async () => {
    const { createGoal } = await import('./goals.js');
    apiRequest.mockResolvedValue({});

    await createGoal({ cookies }, { name: 'Q3', type_weights: { RENEWAL: 0.5 } });

    expect(apiRequest.mock.calls[0][1].body.type_weights).toEqual({ RENEWAL: 0.5 });
  });

  it('carries the stored weights through to the list', async () => {
    respond({ goals: [goal({ type_weights: { RENEWAL: 0.5 } })] });

    const { goals } = await listGoals({ cookies });

    expect(goals[0].type_weights).toEqual({ RENEWAL: 0.5 });
  });

  it('reports an unweighted goal as an empty map rather than undefined', async () => {
    respond({ goals: [goal()] });

    const { goals } = await listGoals({ cookies });

    expect(goals[0].type_weights).toEqual({});
  });
});

describe('listGoals filters', () => {
  beforeEach(() => {
    apiRequest.mockReset();
  });

  it('asks the API for the filters the page was given', async () => {
    respond({ goals: [] });

    await listGoals({ cookies, url: new URL('http://x/goals?period_type=QUARTERLY&q=emea') });

    const listUrl = apiRequest.mock.calls.map((c) => c[0]).find((u) => !u.includes('leaderboard'));
    expect(listUrl).toContain('period_type=QUARTERLY');
    expect(listUrl).toContain('search=emea');
  });

  it('sends only current goals when the page asks for them', async () => {
    respond({ goals: [] });

    await listGoals({ cookies, url: new URL('http://x/goals?window=current') });

    const listUrl = apiRequest.mock.calls.map((c) => c[0]).find((u) => !u.includes('leaderboard'));
    expect(listUrl).toContain('current=true');
  });

  it('ignores a period type that is not one the backend accepts', async () => {
    respond({ goals: [] });

    await listGoals({ cookies, url: new URL('http://x/goals?period_type=; DROP TABLE') });

    const listUrl = apiRequest.mock.calls.map((c) => c[0]).find((u) => !u.includes('leaderboard'));
    expect(listUrl).not.toContain('period_type');
  });

  it('reports the active filters back to the page', async () => {
    respond({ goals: [] });

    const data = await listGoals({
      cookies,
      url: new URL('http://x/goals?period_type=YEARLY&q=emea&window=current')
    });

    expect(data.filters).toEqual({ period_type: 'YEARLY', q: 'emea', window: 'current' });
  });
});

describe('getGoalHistory', () => {
  beforeEach(() => {
    apiRequest.mockReset();
  });

  it('reads finished periods from the history endpoint', async () => {
    apiRequest.mockResolvedValue({
      history: [
        {
          period_start: '2026-01-01',
          period_end: '2026-01-31',
          period_type: 'MONTHLY',
          goal_type: 'REVENUE',
          goals_count: 2,
          attained_count: 1,
          target: 300,
          achieved: 240,
          percent: 80,
          goals: [goal({ name: 'Jan' })]
        }
      ]
    });

    const { getGoalHistory } = await import('./goals.js');
    const { history } = await getGoalHistory({ cookies });

    expect(apiRequest.mock.calls[0][0]).toContain('/opportunities/goals/history/');
    expect(history[0].percent).toBe(80);
    expect(history[0].goal_type).toBe('REVENUE');
    expect(history[0].attained_count).toBe(1);
    expect(history[0].goals[0].name).toBe('Jan');
  });

  it('returns an empty history rather than throwing when there is none', async () => {
    apiRequest.mockResolvedValue({});

    const { getGoalHistory } = await import('./goals.js');

    expect((await getGoalHistory({ cookies })).history).toEqual([]);
  });
});

describe('getCurrentGoals', () => {
  beforeEach(() => {
    apiRequest.mockReset();
  });

  it('asks only for goals running today that are not paused', async () => {
    apiRequest.mockResolvedValue({ goals: [] });
    const { getCurrentGoals } = await import('./goals.js');

    await getCurrentGoals({ cookies });

    const url = apiRequest.mock.calls[0][0];
    expect(url).toContain('current=true');
    expect(url).toContain('active=true');
  });

  it('maps the rows the same way the goals list does', async () => {
    apiRequest.mockResolvedValue({ goals: [goal({ name: 'Q3', progress_percent: 62 })] });
    const { getCurrentGoals } = await import('./goals.js');

    const goals = await getCurrentGoals({ cookies });

    expect(goals[0].name).toBe('Q3');
    expect(goals[0].progress_percent).toBe(62);
    expect(goals[0].type_weights).toEqual({});
  });

  it('returns nothing rather than breaking the dashboard when the call fails', async () => {
    apiRequest.mockRejectedValue(new Error('boom'));
    const { getCurrentGoals } = await import('./goals.js');

    expect(await getCurrentGoals({ cookies })).toEqual([]);
  });
});

File: mobile/test/unit/goals_test.dart
import 'package:bottle_crm/data/models/sales_goal.dart';
import 'package:bottle_crm/providers/goals_provider.dart';
import 'package:bottle_crm/screens/goals/goals_screen.dart';
import 'package:flutter_test/flutter_test.dart';

SalesGoal goal({
  String id = 'g1',
  String name = 'Goal',
  String status = 'on_track',
  bool isActive = true,
  double target = 1000,
  double progress = 100,
  int percent = 10,
  String periodEnd = '2026-12-31',
  String? assignedToId,
  String? assignedToName,
  String? teamId,
  String? teamName,
  String goalType = 'REVENUE',
}) => SalesGoal(
  id: id,
  name: name,
  goalType: goalType,
  targetValue: target,
  periodType: 'MONTHLY',
  periodStart: '2026-01-01',
  periodEnd: periodEnd,
  assignedToId: assignedToId,
  assignedToName: assignedToName,
  teamId: teamId,
  teamName: teamName,
  isActive: isActive,
  progressValue: progress,
  progressPercent: percent,
  status: status,
);

void main() {
  group('SalesGoal.fromJson', () {
    test('flattens the nested assignee and team the API sends', () {
      final parsed = SalesGoal.fromJson({
        'id': 'g1',
        'name': 'Q1 revenue',
        'goal_type': 'REVENUE',
        'target_value': '50000.00',
        'period_type': 'QUARTERLY',
        'period_start': '2026-01-01',
        'period_end': '2026-03-31',
        'assigned_to': 'p1',
        'assigned_to_detail': {
          'id': 'p1',
          'user_details': {'name': 'Ada Lovelace', 'email': 'ada@example.com'},
        },
        'team': null,
        'team_detail': null,
        'is_active': true,
        'progress_value': 20000,
        'progress_percent': 40,
        'status': 'behind',
      });

      expect(parsed.assignedToId, 'p1');
      expect(parsed.assignedToName, 'Ada Lovelace');
      // `target_value` is a DecimalField, so it arrives as a string.
      expect(parsed.targetValue, 50000);
      expect(parsed.progressPercent, 40);
    });

    test('falls back to the email when a person has no name', () {
      final parsed = SalesGoal.fromJson({
        'id': 'g1',
        'assigned_to_detail': {
          'id': 'p1',
          'user_details': {'name': '', 'email': 'ada@example.com'},
        },
      });
      expect(parsed.assignedToName, 'ada@example.com');
    });

    test('survives a payload with nothing in it', () {
      final parsed = SalesGoal.fromJson(const {});
      expect(parsed.id, '');
      expect(parsed.targetValue, 0);
      expect(parsed.assignedToName, isNull);
    });
  });

  group('targetLabel', () {
    test('names the person, then the team, then the org', () {
      expect(
        goal(assignedToId: 'p1', assignedToName: 'Ada').targetLabel,
        'Ada',
      );
      expect(
        goal(teamId: 't1', teamName: 'Support').targetLabel,
        'Support (team)',
      );
      expect(goal().targetLabel, 'Whole organisation');
    });

    test('prefers the person when the API sends both', () {
      // The API permits both FKs. Neither client can write that state, but a
      // row created by curl can arrive in it, and the label must pick one.
      final both = goal(
        assignedToId: 'p1',
        assignedToName: 'Ada',
        teamId: 't1',
        teamName: 'Support',
      );
      expect(both.targetLabel, 'Ada');
    });
  });

  group('goalTotals', () {
    test('sums the active goals and counts every one', () {
      final totals = goalTotals([
        goal(id: 'a', target: 100, progress: 50),
        goal(id: 'b', target: 200, progress: 20),
        goal(id: 'c', target: 999, progress: 999, isActive: false),
      ], today: '2026-06-01');

      expect(totals.count, 3);
      expect(totals.active, 2);
      // The retired goal's 999 is in neither sum.
      expect(totals.target, 300);
      expect(totals.achieved, 70);
    });

    test('counts a goal ending today as still behind pace', () {
      // The boundary the web got wrong: it compared `period_end` as a UTC
      // instant against the clock, so a goal dropped out of this count part-way
      // through its own final day. A goal ending today is one somebody can
      // still act on.
      final totals = goalTotals([
        goal(status: 'behind', periodEnd: '2026-06-01'),
      ], today: '2026-06-01');
      expect(totals.behind, 1);
    });

    test('leaves out a goal whose period has already ended', () {
      final totals = goalTotals([
        goal(status: 'behind', periodEnd: '2026-05-31'),
      ], today: '2026-06-01');
      expect(totals.behind, 0);
    });

    test('leaves out a retired goal even when it is behind and open', () {
      final totals = goalTotals([
        goal(status: 'behind', periodEnd: '2026-12-31', isActive: false),
      ], today: '2026-06-01');
      expect(totals.behind, 0);
    });

    test('is all zeroes for no goals rather than throwing', () {
      final totals = goalTotals(const [], today: '2026-06-01');
      expect(totals.count, 0);
      expect(totals.target, 0);
    });
  });

  group('goalToday', () {
    test('is the local date, zero-padded', () {
      expect(goalToday(DateTime(2026, 1, 5)), '2026-01-05');
      expect(goalToday(DateTime(2026, 12, 31)), '2026-12-31');
    });

    test('does not shift the day for a late-evening local time', () {
      // `toIso8601String()` on a UTC conversion would hand back the next day
      // here for anywhere west of Greenwich, and the previous one east of it.
      // These are date-only fields, so a day is the whole quantity.
      expect(goalToday(DateTime(2026, 6, 1, 23, 59)), '2026-06-01');
      expect(goalToday(DateTime(2026, 6, 1, 0, 1)), '2026-06-01');
    });
  });

  group('sortGoalsByUrgency', () {
    test('puts active before retired, whatever the status', () {
      final sorted = sortGoalsByUrgency([
        goal(id: 'retired', status: 'behind', isActive: false),
        goal(id: 'active', status: 'completed'),
      ]);
      expect(sorted.map((g) => g.id), ['active', 'retired']);
    });

    test('orders behind, at risk, on track, then met', () {
      final sorted = sortGoalsByUrgency([
        goal(id: 'done', status: 'completed'),
        goal(id: 'ok', status: 'on_track'),
        goal(id: 'bad', status: 'behind'),
        goal(id: 'risky', status: 'at_risk'),
      ]);
      expect(sorted.map((g) => g.id), ['bad', 'risky', 'ok', 'done']);
    });

    test('breaks a tie on progress, furthest along first', () {
      final sorted = sortGoalsByUrgency([
        goal(id: 'low', status: 'behind', percent: 10),
        goal(id: 'high', status: 'behind', percent: 40),
      ]);
      expect(sorted.map((g) => g.id), ['high', 'low']);
    });

    test('does not mutate what it was given', () {
      final input = [
        goal(id: 'done', status: 'completed'),
        goal(id: 'bad', status: 'behind'),
      ];
      sortGoalsByUrgency(input);
      expect(input.map((g) => g.id), ['done', 'bad']);
    });
  });

  group('validateGoalForm', () {
    String? check({
      String name = 'Goal',
      String target = '100',
      String start = '2026-01-01',
      String end = '2026-03-31',
    }) => validateGoalForm(
      name: name,
      targetValue: target,
      periodStart: start,
      periodEnd: end,
    );

    test('accepts a filled-in form', () {
      expect(check(), isNull);
    });

    test('refuses a blank name', () {
      expect(check(name: '   '), isNotNull);
    });

    test('refuses a target that is not a number', () {
      expect(check(target: 'lots'), contains('number'));
    });

    test('refuses zero and negative targets, matching the serializer', () {
      // `SalesGoalCreateSerializer.validate` rejects `target_value <= 0`.
      expect(check(target: '0'), contains('greater than 0'));
      expect(check(target: '-5'), contains('greater than 0'));
    });

    test('refuses an end date on or before the start', () {
      // The serializer rejects `period_end <= period_start`, so equal dates are
      // refused too and the message must not imply a one-day goal is possible.
      expect(check(start: '2026-03-31', end: '2026-01-01'), isNotNull);
      expect(check(start: '2026-01-01', end: '2026-01-01'), isNotNull);
    });

    test('refuses a missing date rather than sending a blank one', () {
      expect(check(start: ''), isNotNull);
      expect(check(end: ''), isNotNull);
    });
  });

  group('goalTargetFields', () {
    test('sets exactly one FK and explicitly nulls the other', () {
      // Both keys are always present. PUT is partial, so omitting one keeps the
      // old value, and switching a goal from a person to a team would leave it
      // assigned to both.
      expect(goalTargetFields('profile:p1'), {
        'assigned_to': 'p1',
        'team': null,
      });
      expect(goalTargetFields('team:t1'), {'assigned_to': null, 'team': 't1'});
    });

    test('clears both for a whole-org goal, and for anything unrecognised', () {
      expect(goalTargetFields('org'), {'assigned_to': null, 'team': null});
      expect(goalTargetFields(null), {'assigned_to': null, 'team': null});
      expect(goalTargetFields('nonsense'), {'assigned_to': null, 'team': null});
    });
  });

  group('goalTargetValue', () {
    test('round-trips through goalTargetFields', () {
      expect(goalTargetFields(goalTargetValue(goal(assignedToId: 'p1'))), {
        'assigned_to': 'p1',
        'team': null,
      });
      expect(goalTargetFields(goalTargetValue(goal(teamId: 't1'))), {
        'assigned_to': null,
        'team': 't1',
      });
      expect(goalTargetFields(goalTargetValue(goal())), {
        'assigned_to': null,
        'team': null,
      });
    });

    test('reads an empty string FK as no target, not as an id', () {
      expect(goalTargetValue(goal(assignedToId: '', teamId: '')), 'org');
    });
  });

  group('GoalLeaderRow.fromJson', () {
    test('reads the name the endpoint now sends', () {
      final row = GoalLeaderRow.fromJson({
        'rank': 1,
        'goal_id': 'g1',
        'goal_name': 'Q1',
        'user': {'id': 'p1', 'name': 'Ada Lovelace'},
        'target': 100.0,
        'achieved': 104.0,
        'percent': 104,
      });
      expect(row.user, 'Ada Lovelace');
      // Uncapped, unlike SalesGoal.progressPercent. 104% is the point of a board.
      expect(row.percent, 104);
    });

    test('says Unknown rather than blank when the user block is missing', () {
      expect(GoalLeaderRow.fromJson(const {'rank': 1}).user, 'Unknown');
    });
  });

  group('labels', () {
    test('spell each status for a person', () {
      expect(goalStatusLabel('behind'), 'Behind pace');
      expect(goalStatusLabel('at_risk'), 'At risk');
      expect(goalStatusLabel('on_track'), 'On track');
      expect(goalStatusLabel('completed'), 'Target met');
    });

    test('pass an unrecognised status through rather than blanking it', () {
      expect(goalStatusLabel('surprise'), 'surprise');
    });

    test('cover every type and period the backend accepts', () {
      for (final type in goalTypes) {
        expect(goalTypeLabel(type), isNot(type));
      }
      for (final period in goalPeriodTypes) {
        expect(goalPeriodLabel(period), isNotEmpty);
      }
    });
  });

  _activitiesAndWeights();
  _dateRangeTests();
  _filterTests();
}

void _activitiesAndWeights() {
  group('activity goals', () {
    test('offers ACTIVITIES as a goal type the backend accepts', () {
      expect(goalTypes, contains('ACTIVITIES'));
    });

    test('labels an activities goal', () {
      expect(goalTypeLabel('ACTIVITIES'), 'Activities');
    });

    test('counts activities rather than formatting them as money', () {
      // `isMoney` used to read `goalType != 'DEALS_CLOSED'`, so a third goal
      // type would have printed "$40" for a quota of forty logged activities.
      expect(goal(goalType: 'ACTIVITIES').isMoney, isFalse);
      expect(goal(goalType: 'DEALS_CLOSED').isMoney, isFalse);
      expect(goal(goalType: 'REVENUE').isMoney, isTrue);
    });
  });

  group('deal type weights', () {
    test('names the five deal types a weight can be set against', () {
      expect(dealTypes, [
        'NEW_BUSINESS',
        'EXISTING_BUSINESS',
        'RENEWAL',
        'UPSELL',
        'CROSS_SELL',
      ]);
    });

    test('reads the stored weight map off the payload', () {
      final parsed = SalesGoal.fromJson({
        'id': 'g1',
        'name': 'Weighted',
        'goal_type': 'REVENUE',
        'target_value': '100',
        'type_weights': {'RENEWAL': 0.5},
      });

      expect(parsed.typeWeights, {'RENEWAL': 0.5});
    });

    test('reports an unweighted goal as an empty map, never null', () {
      final parsed = SalesGoal.fromJson({'id': 'g1', 'name': 'Plain'});

      expect(parsed.typeWeights, isEmpty);
    });

    test('counts only the types actually re-weighed', () {
      expect(goal().weightedTypeCount, 0);
      expect(
        SalesGoal.fromJson({
          'id': 'g',
          'name': 'n',
          // A weight of exactly 1 is the default, so it is not an adjustment.
          'type_weights': {'RENEWAL': 0.5, 'UPSELL': 1},
        }).weightedTypeCount,
        1,
      );
    });
  });

  group('formatGoalValue', () {
    test('counts deals as deals', () {
      expect(
        formatGoalValue(3, goalType: 'DEALS_CLOSED', symbol: r'$'),
        '3 deals',
      );
      expect(
        formatGoalValue(1, goalType: 'DEALS_CLOSED', symbol: r'$'),
        '1 deal',
      );
    });

    test('counts activities as activities, not as deals', () {
      // The unit used to hang off `isMoney`, so anything that was not revenue
      // was printed as deals: a quota of forty logged activities read "40
      // deals".
      expect(
        formatGoalValue(40, goalType: 'ACTIVITIES', symbol: r'$'),
        '40 activities',
      );
      expect(
        formatGoalValue(1, goalType: 'ACTIVITIES', symbol: r'$'),
        '1 activity',
      );
    });

    test('prices revenue in the org currency', () {
      expect(
        formatGoalValue(120000, goalType: 'REVENUE', symbol: r'$'),
        contains(r'$'),
      );
    });
  });

  group('GoalHistoryPeriod', () {
    test('reads a finished period and keeps its goal type', () {
      final period = GoalHistoryPeriod.fromJson({
        'period_start': '2026-01-01',
        'period_end': '2026-01-31',
        'period_type': 'MONTHLY',
        'goal_type': 'REVENUE',
        'goals_count': 2,
        'attained_count': 1,
        'target': 300,
        'achieved': 240,
        'percent': 80,
        'goals': [
          {'id': 'g1', 'name': 'Jan', 'goal_type': 'REVENUE'},
        ],
      });

      expect(period.goalType, 'REVENUE');
      expect(period.percent, 80);
      expect(period.attainedCount, 1);
      expect(period.goals.single.name, 'Jan');
      expect(period.isMoney, isTrue);
    });

    test('does not format a deals period as money', () {
      final period = GoalHistoryPeriod.fromJson({
        'goal_type': 'DEALS_CLOSED',
        'target': 56,
        'achieved': 4,
        'percent': 7,
      });

      expect(period.isMoney, isFalse);
    });
  });
}

void _dateRangeTests() {
  group('goalDateRange', () {
    test('prints a period as two short dates', () {
      expect(goalDateRange('2026-05-01', '2026-05-31'), '1 May - 31 May');
    });

    test('hands back a malformed value rather than inventing a date', () {
      expect(
        goalDateRange('not-a-date', '2026-05-31'),
        'not-a-date to 2026-05-31',
      );
    });
  });
}

void _filterTests() {
  group('goalListQuery', () {
    test('asks for everything when nothing is filtered', () {
      final query = goalListQuery(const GoalFilters());
      expect(query, contains('limit=1000'));
      expect(query, isNot(contains('search=')));
      expect(query, isNot(contains('period_type=')));
    });

    test('sends the search text and the period', () {
      final query = goalListQuery(
        const GoalFilters(query: 'emea', periodType: 'QUARTERLY'),
      );
      expect(query, contains('search=emea'));
      expect(query, contains('period_type=QUARTERLY'));
    });

    test('encodes search text rather than pasting it into the URL', () {
      final query = goalListQuery(const GoalFilters(query: 'a&b c'));
      expect(query, isNot(contains('a&b c')));
      expect(query, contains('search=a%26b+c'));
    });

    test('maps the window onto the flag the API actually takes', () {
      expect(
        goalListQuery(const GoalFilters(window: 'current')),
        contains('current=true'),
      );
      expect(
        goalListQuery(const GoalFilters(window: 'active')),
        contains('active=true'),
      );
    });

    test('drops a period the backend does not accept', () {
      // The value reaches this from a picker today, but a filter that silently
      // forwarded anything would turn a future typo into an empty list with no
      // explanation.
      expect(
        goalListQuery(const GoalFilters(periodType: 'FORTNIGHTLY')),
        isNot(contains('period_type')),
      );
    });

    test('knows when a filter is on, so the empty state can say so', () {
      expect(const GoalFilters().isFiltered, isFalse);
      expect(const GoalFilters(query: 'x').isFiltered, isTrue);
      expect(const GoalFilters(window: 'current').isFiltered, isTrue);
      expect(const GoalFilters(periodType: 'MONTHLY').isFiltered, isTrue);
    });
  });
}
